File: scripts/analyze_robustness.py
#!/usr/bin/env python3
"""
Analyze robustness sweep results and generate publication-ready summary.
"""

import pickle
import pandas as pd
import numpy as np
from pathlib import Path
from scipy import stats


def analyze(results_base: str = "glimpse_robustness_sweep"):
    """Analyze all robustness configurations and generate summary tables."""

    results_dir = Path(results_base)

    if not results_dir.exists():
        print(f"Error: Results directory '{results_dir}' not found.")
        return

    configs = [d for d in results_dir.iterdir() if d.is_dir()]

    if not configs:
        print(f"Error: No configuration directories found in '{results_dir}'")
        return

    print("=" * 80)
    print("ROBUSTNESS ANALYSIS: Fixed-Tier Causal Effects Across Parameter Space")
    print("=" * 80)
    print(f"\nFound {len(configs)} configurations to analyze.")

    # Collect results from all configurations
    all_results = []

    for config_dir in sorted(configs):
        config_name = config_dir.name

        # Load agent data from all runs
        run_dirs = list(config_dir.glob('Fixed_AI_Level_*'))

        if not run_dirs:
            print(f"  Warning: No runs found in {config_name}")
            continue

        agents_list = []
        for run_dir in run_dirs:
            agents_file = run_dir / 'final_agents.pkl'
            if agents_file.exists():
                with open(agents_file, 'rb') as f:
                    agents_df = pickle.load(f)
                tier = run_dir.name.split('_')[3]
                agents_df['ai_tier'] = tier
                agents_df['config'] = config_name
                agents_list.append(agents_df)

        if not agents_list:
            continue

        agents_df = pd.concat(agents_list, ignore_index=True)

        # Compute survival by tier
        baseline = agents_df[agents_df['ai_tier'] == 'none']['survived']
        baseline_mean = baseline.mean()
        baseline_std = baseline.std()

        for tier in ['basic', 'advanced', 'premium']:
            treatment = agents_df[agents_df['ai_tier'] == tier]['survived']

            if len(treatment) == 0:
                continue

            ate = treatment.mean() - baseline_mean
            pooled_std = np.sqrt((baseline_std**2 + treatment.std()**2) / 2)
            cohens_d = ate / pooled_std if pooled_std > 0 else 0

            t_stat, p_value = stats.ttest_ind(treatment, baseline)
            se = np.sqrt(baseline.var()/len(baseline) + treatment.var()/len(treatment))
            ci_lower = ate - 1.96 * se
            ci_upper = ate + 1.96 * se

            all_results.append({
                'config': config_name,
                'comparison': f'{tier}_vs_none',
                'baseline_survival': baseline_mean,
                'treatment_survival': treatment.mean(),
                'ate': ate,
                'ci_lower': ci_lower,
                'ci_upper': ci_upper,
                'cohens_d': cohens_d,
                'p_value': p_value,
                'n_treatment': len(treatment),
                'n_control': len(baseline),
            })

        print(f"  Processed: {config_name} ({len(run_dirs)} runs)")

    if not all_results:
        print("Error: No results to analyze.")
        return

    results_df = pd.DataFrame(all_results)

    # Print summary table
    print("\n" + "=" * 80)
    print("TABLE: Robustness of Causal Effects Across Parameter Configurations")
    print("=" * 80)

    # Pivot for display
    summary = results_df.pivot_table(
        index='config',
        columns='comparison',
        values=['ate', 'cohens_d'],
        aggfunc='first'
    )

    print("\nAverage Treatment Effects (ATE):")
    print("-" * 60)
    ate_table = results_df.pivot(index='config', columns='comparison', values='ate')
    print(ate_table.round(3).to_string())

    print("\n\nCohen's d Effect Sizes:")
    print("-" * 60)
    d_table = results_df.pivot(index='config', columns='comparison', values='cohens_d')
    print(d_table.round(3).to_string())

    # Robustness bounds
    print("\n" + "=" * 80)
    print("ROBUSTNESS BOUNDS (Range of Effects Across All Configurations)")
    print("=" * 80)

    for comparison in ['basic_vs_none', 'advanced_vs_none', 'premium_vs_none']:
        subset = results_df[results_df['comparison'] == comparison]

        print(f"\n{comparison}:")
        print(f"  ATE Range:      [{subset['ate'].min():.3f}, {subset['ate'].max():.3f}]")
        print(f"  ATE Mean ± SD:  {subset['ate'].mean():.3f} ± {subset['ate'].std():.3f}")
        print(f"  Cohen's d Range: [{subset['cohens_d'].min():.3f}, {subset['cohens_d'].max():.3f}]")

        # Check if effect direction is consistent
        all_negative = (subset['ate'] < 0).all()
        print(f"  Direction consistent: {'YES (all negative)' if all_negative else 'NO (some positive)'}")

    # Save results
    output_file = results_dir / 'robustness_summary.csv'
    results_df.to_csv(output_file, index=False)
    print(f"\n\nResults saved to: {output_file}")

    # Create publication table
    pub_table = []
    for config in results_df['config'].unique():
        config_data = results_df[results_df['config'] == config]
        row = {'Configuration': config}
        for _, r in config_data.iterrows():
            comp = r['comparison'].replace('_vs_none', '')
            row[f'{comp}_ATE'] = f"{r['ate']:.3f}"
            row[f'{comp}_d'] = f"{r['cohens_d']:.2f}"
        pub_table.append(row)

    pub_df = pd.DataFrame(pub_table)
    pub_file = results_dir / 'robustness_publication_table.csv'
    pub_df.to_csv(pub_file, index=False)
    print(f"Publication table saved to: {pub_file}")

    return results_df


if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1:
        analyze(sys.argv[1])
    else:
        analyze()

File: julia/tools/analyze_robustness.jl
"""
Analyze robustness sweep results and generate publication-ready summary.

Port of: glimpse_abm/scripts/analyze_robustness.py
"""

module AnalyzeRobustness

using DataFrames
using Statistics
using Arrow
using Serialization

export analyze_robustness, generate_robustness_report

"""
Load agent data from a run directory.
"""
function load_agents_from_run(run_dir::String)::Union{DataFrame,Nothing}
    # Try different file formats
    for (filename, loader) in [
        ("final_agents.arrow", path -> DataFrame(Arrow.Table(path))),
        ("final_agents.jld2", path -> begin
            jld = JLD2.jldopen(path, "r")
            df = jld["agents"]
            close(jld)
            df
        end)
    ]
        path = joinpath(run_dir, filename)
        if isfile(path)
            try
                return loader(path)
            catch e
                continue
            end
        end
    end
    return nothing
end

"""
Calculate Cohen's d effect size.
"""
function cohens_d(group1::Vector{Float64}, group2::Vector{Float64})::Float64
    n1, n2 = length(group1), length(group2)
    if n1 < 2 || n2 < 2
        return 0.0
    end

    v1, v2 = var(group1), var(group2)
    pooled_std = sqrt(((n1 - 1) * v1 + (n2 - 1) * v2) / (n1 + n2 - 2))

    if pooled_std ≈ 0
        return 0.0
    end

    return (mean(group1) - mean(group2)) / pooled_std
end

"""
Two-sample t-test (Welch's t-test).
"""
function welch_ttest(group1::Vector{Float64}, group2::Vector{Float64})::Tuple{Float64,Float64}
    n1, n2 = length(group1), length(group2)
    if n1 < 2 || n2 < 2
        return (NaN, NaN)
    end

    m1, m2 = mean(group1), mean(group2)
    v1, v2 = var(group1), var(group2)

    se = sqrt(v1/n1 + v2/n2)
    if se ≈ 0
        return (0.0, 1.0)
    end

    t_stat = (m1 - m2) / se

    # Welch-Satterthwaite degrees of freedom
    df = (v1/n1 + v2/n2)^2 / ((v1/n1)^2/(n1-1) + (v2/n2)^2/(n2-1))

    # Approximate p-value using normal distribution (good for large df)
    p_value = 2 * (1 - cdf(Normal(), abs(t_stat)))

    return (t_stat, p_value)
end

using Distributions: Normal, cdf

"""
Analyze all robustness configurations and generate summary tables.
"""
function analyze_robustness(results_base::String="glimpse_robustness_sweep")::Union{DataFrame,Nothing}
    results_dir = results_base

    if !isdir(results_dir)
        println("Error: Results directory '$results_dir' not found.")
        return nothing
    end

    configs = [d for d in readdir(results_dir, join=true) if isdir(d)]

    if isempty(configs)
        println("Error: No configuration directories found in '$results_dir'")
        return nothing
    end

    println("=" ^ 80)
    println("ROBUSTNESS ANALYSIS: Fixed-Tier Causal Effects Across Parameter Space")
    println("=" ^ 80)
    println("\nFound $(length(configs)) configurations to analyze.")

    # Collect results from all configurations
    all_results = Dict{String,Any}[]

    for config_dir in sort(configs)
        config_name = basename(config_dir)

        # Load agent data from all runs
        run_dirs = [d for d in readdir(config_dir, join=true)
                    if isdir(d) && startswith(basename(d), "Fixed_AI_Level_")]

        if isempty(run_dirs)
            println("  Warning: No runs found in $config_name")
            continue
        end

        agents_list = DataFrame[]
        for run_dir in run_dirs
            agents_df = load_agents_from_run(run_dir)
            if agents_df !== nothing
                tier = split(basename(run_dir), "_")[4]
                agents_df[!, :ai_tier] .= tier
                agents_df[!, :config] .= config_name
                push!(agents_list, agents_df)
            end
        end

        if isempty(agents_list)
            continue
        end

        agents_df = vcat(agents_list...)

        # Ensure survived column exists and is numeric
        if !("survived" in names(agents_df))
            if "final_status" in names(agents_df)
                agents_df[!, :survived] = agents_df.final_status .== "active"
            elseif "alive" in names(agents_df)
                agents_df[!, :survived] = agents_df.alive
            else
                continue
            end
        end

        # Get baseline (no AI)
        baseline_mask = agents_df.ai_tier .== "none"
        baseline = Float64.(coalesce.(agents_df[baseline_mask, :survived], false))

        if isempty(baseline)
            continue
        end

        baseline_mean = mean(baseline)
        baseline_std = std(baseline)

        # Calculate effects for each AI tier
        for tier in ["basic", "advanced", "premium"]
            treatment_mask = agents_df.ai_tier .== tier
            treatment = Float64.(coalesce.(agents_df[treatment_mask, :survived], false))

            if isempty(treatment)
                continue
            end

            # Average Treatment Effect
            ate = mean(treatment) - baseline_mean

            # Cohen's d
            pooled_std = sqrt((baseline_std^2 + std(treatment)^2) / 2)
            d = pooled_std > 0 ? ate / pooled_std : 0.0

            # T-test for p-value
            _, p_value = welch_ttest(treatment, baseline)

            # Confidence interval
            se = sqrt(var(baseline)/length(baseline) + var(treatment)/length(treatment))
            ci_lower = ate - 1.96 * se
            ci_upper = ate + 1.96 * se

            push!(all_results, Dict{String,Any}(
                "config" => config_name,
                "comparison" => "$(tier)_vs_none",
                "baseline_survival" => baseline_mean,
                "treatment_survival" => mean(treatment),
                "ate" => ate,
                "ci_lower" => ci_lower,
                "ci_upper" => ci_upper,
                "cohens_d" => d,
                "p_value" => p_value,
                "n_treatment" => length(treatment),
                "n_control" => length(baseline)
            ))
        end

        println("  Processed: $config_name ($(length(run_dirs)) runs)")
    end

    if isempty(all_results)
        println("Error: No results to analyze.")
        return nothing
    end

    results_df = DataFrame(all_results)

    # Print summary table
    println("\n" * "=" ^ 80)
    println("TABLE: Robustness of Causal Effects Across Parameter Configurations")
    println("=" ^ 80)

    # Summary by comparison
    println("\nAverage Treatment Effects (ATE):")
    println("-" ^ 60)

    for comp in unique(results_df.comparison)
        subset = results_df[results_df.comparison .== comp, :]
        println("$comp:")
        println("  Mean ATE: $(round(mean(subset.ate), digits=4))")
        println("  ATE Range: [$(round(minimum(subset.ate), digits=4)), $(round(maximum(subset.ate), digits=4))]")
    end

    println("\n\nCohen's d Effect Sizes:")
    println("-" ^ 60)

    for comp in unique(results_df.comparison)
        subset = results_df[results_df.comparison .== comp, :]
        println("$comp:")
        println("  Mean d: $(round(mean(subset.cohens_d), digits=3))")
        println("  d Range: [$(round(minimum(subset.cohens_d), digits=3)), $(round(maximum(subset.cohens_d), digits=3))]")
    end

    # Robustness bounds
    println("\n" * "=" ^ 80)
    println("ROBUSTNESS BOUNDS (Range of Effects Across All Configurations)")
    println("=" ^ 80)

    for comparison in ["basic_vs_none", "advanced_vs_none", "premium_vs_none"]
        subset = results_df[results_df.comparison .== comparison, :]

        if isempty(subset)
            continue
        end

        println("\n$comparison:")
        println("  ATE Range:      [$(round(minimum(subset.ate), digits=3)), $(round(maximum(subset.ate), digits=3))]")
        println("  ATE Mean ± SD:  $(round(mean(subset.ate), digits=3)) ± $(round(std(subset.ate), digits=3))")
        println("  Cohen's d Range: [$(round(minimum(subset.cohens_d), digits=3)), $(round(maximum(subset.cohens_d), digits=3))]")

        # Check if effect direction is consistent
        all_negative = all(subset.ate .< 0)
        println("  Direction consistent: $(all_negative ? "YES (all negative)" : "NO (some positive)")")
    end

    # Save results
    output_file = joinpath(results_dir, "robustness_summary.csv")
    # Using simple CSV writing
    open(output_file, "w") do f
        # Header
        cols = names(results_df)
        println(f, join(cols, ","))
        # Data
        for row in eachrow(results_df)
            values = [string(row[c]) for c in cols]
            println(f, join(values, ","))
        end
    end
    println("\n\nResults saved to: $output_file")

    # Create publication table
    pub_table = Dict{String,Any}[]
    for config in unique(results_df.config)
        config_data = results_df[results_df.config .== config, :]
        row = Dict{String,Any}("Configuration" => config)
        for r in eachrow(config_data)
            comp = replace(r.comparison, "_vs_none" => "")
            row["$(comp)_ATE"] = round(r.ate, digits=3)
            row["$(comp)_d"] = round(r.cohens_d, digits=2)
        end
        push!(pub_table, row)
    end

    pub_df = DataFrame(pub_table)
    pub_file = joinpath(results_dir, "robustness_publication_table.csv")
    open(pub_file, "w") do f
        cols = names(pub_df)
        println(f, join(cols, ","))
        for row in eachrow(pub_df)
            values = [string(row[c]) for c in cols]
            println(f, join(values, ","))
        end
    end
    println("Publication table saved to: $pub_file")

    return results_df
end

"""
Generate a summary report string.
"""
function generate_robustness_report(results_df::DataFrame)::String
    lines = String[]

    push!(lines, "=" ^ 60)
    push!(lines, "ROBUSTNESS ANALYSIS SUMMARY")
    push!(lines, "=" ^ 60)
    push!(lines, "")

    push!(lines, "Configurations analyzed: $(length(unique(results_df.config)))")
    push!(lines, "")

    for comparison in ["basic_vs_none", "advanced_vs_none", "premium_vs_none"]
        subset = results_df[results_df.comparison .== comparison, :]
        if isempty(subset)
            continue
        end

        push!(lines, "$comparison:")
        push!(lines, "  ATE: $(round(mean(subset.ate), digits=4)) [$(round(minimum(subset.ate), digits=4)), $(round(maximum(subset.ate), digits=4))]")
        push!(lines, "  Cohen's d: $(round(mean(subset.cohens_d), digits=3))")

        sig_count = count(subset.p_value .< 0.05)
        push!(lines, "  Significant (p<0.05): $sig_count / $(nrow(subset)) configs")
        push!(lines, "")
    end

    return join(lines, "\n")
end

"""
Main entry point for command-line usage.
"""
function main(args=ARGS)
    if isempty(args)
        results_base = "glimpse_robustness_sweep"
    else
        results_base = args[1]
    end

    analyze_robustness(results_base)
end

end # module AnalyzeRobustness
